quinn-proto/src/token.rs

```diff
@@ -50,34 +67,77 @@ impl RetryToken {
         retry_src_cid: &ConnectionId,
         raw_token_bytes: &[u8],
     ) -> Result<Self, TokenDecodeError> {
-        let aead_key = key.aead_from_hkdf(retry_src_cid);
-        let mut sealed_token = raw_token_bytes.to_vec();
-
-        let data = aead_key.open(&mut sealed_token, &[])?;
-        let mut reader = io::Cursor::new(data);
-        let token_addr = decode_addr(&mut reader).ok_or(TokenDecodeError::UnknownToken)?;
-        if token_addr != *address {
-            return Err(TokenDecodeError::WrongAddress);
-        }
-        let orig_dst_cid =
-            ConnectionId::decode_long(&mut reader).ok_or(TokenDecodeError::UnknownToken)?;
-        let issued = UNIX_EPOCH
-            + Duration::new(
-                reader
-                    .get::<u64>()
-                    .map_err(|_| TokenDecodeError::UnknownToken)?,
-                0,
-            );
-
-        Ok(Self {
-            orig_dst_cid,
-            issued,
+        let last_idx = raw_token_bytes
+            .len()
+            .checked_sub(1)
+            .ok_or(TokenDecodeError::InvalidMaybeNewToken)?;
+        Ok(match raw_token_bytes[last_idx] {
```

Comment:
I just did it like that to simplify encoding. `AeadKey::seal` accepts `&mut Vec<u8>` rather than `&mut [u8]`, so by putting the tag at the end the encoder can do this:

```rs
aead_key.seal(&mut buf, &[0]).unwrap();
buf.push(0);
buf
```

Otherwise it'd require a `buf.insert(0, 0)`.

Reply:
That makes sense, but now I notice that the tag itself isn't encrypted or authenticated. I think it probably should be.

Reply:
It is authenticated. Right? That's what the `&[0]` parameter is in the `AeadKey::seal` call. Anyways, this is probably moot as, per our discussion on Discord, we're thinking we'll make both variants use the same key derivation, so I'll switch it so that it's all encrypted.

Reply:
Anyways, I totally reworked how the encryption works, per your suggestion over Discord.